### ISSUE TITLE
Error fluctuations with well-mixed data

### ISSUE THREAD

**Issue by gokceneraslan** [pytorch/pytorch#3354]
I observe weird training error fluctuations when I train the same model with `shuffle=True` compared to `shuffle=False` argument of `DataLoader()`, even though the data is already well-mixed. 

Here is a reproducible example:

```python
import torch
from torch.autograd import Variable
from torch.utils.data import TensorDataset, DataLoader
import numpy as np
import plotnine as p9

num_sample = 1000
num_feat = 10
num_out = 10

np.random.seed(1923)
torch.manual_seed(1923)

X = np.random.normal(0, 0.5, (num_sample, num_feat)).astype(np.float32)
W = np.random.normal(0, 0.5, (num_feat, num_out)).astype(np.float32)
b = np.random.normal(0, 0.5, (1, num_out)).astype(np.float32)
y_mean = np.dot(X, W) + b

Y = np.random.normal(y_mean, 0.5)

model = torch.nn.Sequential(torch.nn.Linear(num_feat, num_out))
loss = torch.nn.MSELoss()

optimizer = torch.optim.RMSprop(model.parameters(), lr=1e-3)

# dataset
full_data = TensorDataset(torch.from_numpy(X).float(),
                          torch.from_numpy(Y).float())

##### train with shuffling

print('\n##  With shuffle\n')
loader = DataLoader(full_data, batch_size=16, shuffle=True)
hist1 = []

# over epochs
for i in range(300):

    # over batches
    for j, (inp, target) in enumerate(loader):
        inp_var = Variable(inp)
        target_var = Variable(target)

        res = model(inp_var)
        l = loss(res, target_var)

        if i % 30 == 0 and j == 0:
            print('Epoch:', i, ' loss: ', l.data[0])

        if j == 0:
            hist1.append(l.data[0])

        optimizer.zero_grad()
        l.backward()
        optimizer.step()

##### now without shuffling
model = torch.nn.Sequential(torch.nn.Linear(num_feat, num_out))
loss = torch.nn.MSELoss()

optimizer = torch.optim.RMSprop(model.parameters(), lr=1e-3)

print('\n##  Without shuffle\n')
loader = DataLoader(full_data, batch_size=16, shuffle=False)
hist2 = []

# over epochs
for i in range(300):

    # over batches
    for j, (inp, target) in enumerate(loader):
        inp_var = Variable(inp)
        target_var = Variable(target)

        res = model(inp_var)
        l = loss(res, target_var)

        if i % 30 == 0 and j == 0:
            print('Epoch:', i, ' loss: ', l.data[0])
        if j == 0:
            hist2.append(l.data[0])

        optimizer.zero_grad()
        l.backward()
        optimizer.step()


(p9.qplot(range(len(hist1)), hist1, xlab=' ', ylab=' ', geom='path', color='"with shuffle"') +
    p9.geom_path(p9.aes(x=range(len(hist2)), y=hist2, color='"w/o shuffle"')) +
    p9.theme_minimal()+
    p9.labs(color='Loss', x='Epoch')).save('shuffle.png')
```

Here is the output:

```

##  With shuffle

Epoch: 0  loss:  1.0320324897766113
Epoch: 30  loss:  0.22753222286701202
Epoch: 60  loss:  0.2699904441833496
Epoch: 90  loss:  0.29355287551879883
Epoch: 120  loss:  0.19778938591480255
Epoch: 150  loss:  0.2268659621477127
Epoch: 180  loss:  0.2604275047779083
Epoch: 210  loss:  0.2825251817703247
Epoch: 240  loss:  0.23147885501384735
Epoch: 270  loss:  0.22723357379436493

##  Without shuffle

Epoch: 0  loss:  1.1800202131271362
Epoch: 30  loss:  0.2247518002986908
Epoch: 60  loss:  0.2239338606595993
Epoch: 90  loss:  0.22393330931663513
Epoch: 120  loss:  0.22393333911895752
Epoch: 150  loss:  0.22393333911895752
Epoch: 180  loss:  0.22393333911895752
Epoch: 210  loss:  0.22393333911895752
Epoch: 240  loss:  0.22393333911895752
Epoch: 270  loss:  0.22393333911895752
```

and the figure:

![shuffle](https://user-images.githubusercontent.com/1140359/32145381-d110ab3c-bcc7-11e7-9d99-fa54e79ca9ab.png)

I guess this is not the intended behaviour, I don't expect to see any difference in well-mixed data. Any ideas?

**Comment by soumith** [comment]
why did you close the issue? can you elaborate? i was looking into it but i can't figure out what's going on yet

**Comment by gokceneraslan** [comment]
I think this fluctuation is just fine, because errors that I save and print are only those from the first batch. As the first batch is random with shuffle=True, the error fluctuates compared to the second case where the model is trained on always the same batches in same order. 

What I was thinking when I opened the issue was a possible mix of input and target order during shuffling, but that's not the case.

**Comment by soumith** [comment]
your dataset is a dummy dataset. it only has randomized samples. whether it is shuffle=True or False the network is seeing random samples. Yet fluctuations are happening only with shuffle=True. that seems counter-intuitive to me. unless you have better insights I'll keep digging further

**Comment by soumith** [comment]
ohhhh now i get it. yea your code sample's error reporting is the reason, thanks for letting me know :)

**Comment by ozancaglayan** [comment]
What is your intuition about the fact that after some 200 epochs, in the non-shuffled case, the loss is not moving at all? I checked the gradients, they are always the same. It's not happening with SGD or if you decrease the learning rate to `1e-4` with rmsprop and Adam. And why it does not get stuck when you shuffle?

**Comment by soumith** [comment]
prob because it converged

**Comment by gokceneraslan** [comment]
Now I feel like there is something weird. I changed my code, so that the training loss is now calculated as the mean of all batch training losses within epoch. Then, I generated another dataset (via negative binomial regression since I'm mostly working with integer count data) and plotted train/val loss with and without shuffling in Torch and Keras. Here is how it looks like:

![shuffle](https://user-images.githubusercontent.com/1140359/32188911-6b711fcc-bda9-11e7-98fd-a51b8983a4d4.png)

Is this fluctuation really normal? @soumith any comments?

Code is here: https://gist.github.com/gokceneraslan/397cf160aed741a9ba2296505026c659

**Comment by gokceneraslan** [comment]
[Here](https://github.com/fchollet/keras/blob/master/keras/callbacks.py#L223) is how keras calculates training loss per epoch, just average over batch losses.  I don't get why pytorch fluctuates so much...

**Comment by gokceneraslan** [comment]
I've simplified the script a bit:

```python
import numpy as np
import pandas as pd
import plotnine as p9

import torch
from torch.autograd import Variable
from torch.utils.data import TensorDataset, DataLoader, Dataset

np.random.seed(555)
torch.manual_seed(555)

num_sample = 300
num_feat = 10
num_out = 10

batch_size = 32

X = np.random.normal(0, 0.5, (num_sample, num_feat)).astype(np.float32)
W = np.random.normal(0, 0.5, (num_feat, num_out)).astype(np.float32)
b = np.random.normal(0, 0.5, (1, num_out)).astype(np.float32)
Y = np.exp(np.dot(X, W) + b)

#### training with shuffle

ds = TensorDataset(torch.from_numpy(X).float(), torch.from_numpy(Y).float())

# train torch models with and without shuffling
model1 = torch.nn.Linear(num_feat, num_out)
opt1 = torch.optim.RMSprop(model1.parameters(), lr=0.01)
loss1 = torch.nn.MSELoss()

train_hist_shuf = []

for epoch in range(300):

    train_batch_losses = []

    # shuffle dataset
    idx = torch.randperm(len(ds))
    ds = TensorDataset(*ds[idx])

    for batch in range(int(np.ceil(len(ds)/batch_size))):
        x, y = ds[batch*batch_size:(batch+1)*batch_size]

        x_var, y_var = Variable(x), Variable(y)
        pred = model1(x_var)
        l = loss1(pred, y_var)
        train_batch_losses.append(l.data[0])

        opt1.zero_grad()
        l.backward()
        opt1.step()

    # save mean of all batch errors within the epoch
    train_hist_shuf.append(np.array(train_batch_losses).mean())

#### now no shuffling

ds = TensorDataset(torch.from_numpy(X).float(), torch.from_numpy(Y).float())

# train torch models with and without shuffling
model2 = torch.nn.Linear(num_feat, num_out)
opt2 = torch.optim.RMSprop(model2.parameters(), lr=0.01)
loss2 = torch.nn.MSELoss()

train_hist_noshuf = []

for epoch in range(300):
    train_batch_losses = []

    for batch in range(int(np.ceil(len(ds)/batch_size))):
        x, y = ds[batch*batch_size:(batch+1)*batch_size]

        x_var, y_var = Variable(x), Variable(y)
        pred = model2(x_var)
        l = loss2(pred, y_var)
        train_batch_losses.append(l.data[0])

        opt2.zero_grad()
        l.backward()
        opt2.step()

    train_hist_noshuf.append(np.array(train_batch_losses).mean())

(p9.ggplot(pd.DataFrame({'torch_train_shuf': train_hist_shuf,
                         'torch_train_noshuf': train_hist_noshuf,
                         'epochs': range(len(train_hist_shuf))}),
            p9.aes(x='epochs')) +
  p9.geom_path(p9.aes(y='torch_train_shuf', color='"loss (shuffled)"')) +
  p9.geom_path(p9.aes(y='torch_train_noshuf', color='"loss (not shuffled)"')) +
  p9.labs(color='Loss', y=' ') +
  p9.theme_minimal()).save('shuffle.png', width=6,
                           height=4, dpi=200)

```

Here is the resulting figure:

![shuffle](https://user-images.githubusercontent.com/1140359/32252805-8f0f6bac-be97-11e7-8f93-7666aa7ecaad.png)

**Comment by soumith** [comment]
This is still expected for the following counter-intuitive reason:

When you do not shuffle across epochs, the model is never seeing the same sample again until it sees all other samples. For example, if you have 100 samples in your dataset, the model sees sample 13 with a frequency of 100, at the 13th step in every epoch.

So, for a randomly distributed dataset, the momentum effect over some sequence of samples is roughly the same.

However, if you shuffle the dataset after every epoch, then there is a higher chance that the frequency of seeing sample 13 is less than 100. i.e. you might see sample 13 in the last iteration of epoch0 and again in the first iteration of epoch1. This will have implications on optimization methods which use gradients and momentum, and the overall training loss might actually take a hit. For example think of a worst-case scenario where: epoch0 - [....., i0, i1, i2, i3], epoch1 - [i3, i2, i1, i0, ...]. In this case, the model can ever so slightly overfit to this local sequence of i0-i3 and when you sample i15 in the next iteration, the loss on it will be higher (because of that temporary overfitting). This happens with pure SGD, but is amplified in momentum-based methods even further.

**Comment by gokceneraslan** [comment]
OK, makes a lot of sense. What puzzles me is that Keras also seems to [shuffle](https://github.com/fchollet/keras/blob/master/keras/engine/training.py#L1194) in every epoch, however I never see these fluctuations in Keras training error i.e. https://github.com/pytorch/pytorch/issues/3354#issuecomment-340544571

**Comment by soumith** [comment]
hmmm, are you using RMSProp for Keras too?
I am not entirely sure why keras would have different behavior.
Can you compare pytorch and keras on SGD with no momentum

**Comment by gokceneraslan** [comment]
Yeah, I used rmsprop with same lr in both pytorch and keras. SGD without momentum is also similar:

![shuffle](https://user-images.githubusercontent.com/1140359/32285360-4e1c9276-bf2a-11e7-9a18-be85f28fa040.png)

I hope it's not a bug in my script :)

**Comment by gokceneraslan** [comment]
Same setup with SGD momentum=0. but this time I passed `drop_last=True` to DataLoader:

![shuffle](https://user-images.githubusercontent.com/1140359/32285852-e6d1b20c-bf2b-11e7-9216-7fb47b99d10e.png)

**Comment by ozancaglayan** [comment]
The situation does not happen if batches have the **same** number of samples but it gets worse when the number of samples in the final batch becomes more and more smaller (i.e. compare batch sizes 100, 101 (91 samples in final batch) and 111 (1 sample in final patch))

**Comment by ozancaglayan** [comment]
To finalize this, we have discovered that the problem was caused by wrong loss calculation which did not take into account that the last batch has less samples than all others.

**Comment by soumith** [comment]
i see. that makes sense too i guess.

**Comment by tkornut** [comment]
I have observed something that might be related to that discussion. I was testing my (frozen) model on a test set and noticed that loss fluctuates. First I thought that if may result from the problem with improper weighting of losses from different batches. So I have set batch size to size of test set. To my surprise, loss was still fluctuating.

So took MNIST dataset, pretrained a shallow model (1 FC+Softmax) and concrete loss (nllloss) values on test set are as follows:
1) when I turn off shuffling ( dataloader ->  shuffle: false) and pass 10k test samples in a single batch:
0.28591325879096985
[the value is "stable" in different runs" - ok]
2) but when pass 10k test samples in one batch with shuffling on the loss is fluctuating, exemplary values:
0.28591349720954895
0.2859136164188385
0.2859131693840027

Additionally, I have left shuffling on, but set torch seed to a constant value. As expected, this resulted in "stable" loss value (same value in n runs):
0.28591397404670715

So what is really going on here? Shouldn't I get EXACTLY the same loss for the same batch, disregarding the order of the samples in that batch?
